sdk/messaging/azservicebus/internal/namespace.go

```diff
@@ -50,6 +50,9 @@ type (
 		clientMu         sync.RWMutex
 		client           *amqp.Client
 		negotiateClaimMu sync.Mutex
+		// indicates that the client was closed permanently, and not just
+		// for recovery.
+		closedPermanently bool
```

Comment:
I have mixed feelings on this.  Shouldn't the owner of ns know if it was closed for recovery vs really closed?

Reply:
The namespace and the client are really 1:1. I just let the namespace hold the flag since it's basically the authority. So it's the same concept, unless I'm missing something.

Reply:
Whenever I see a "really the thing" flag it gives me pause for concern.  Seems like the API is performing double-duty, masquerading as two slightly different APIs.  How does a person know when to write Close(true) or Close(false)?  I know this is all internal stuff so maybe it's a bit academic.

Reply:
It's _very_ internal (and it's a good question):

Ignoring the bool part, the non-permanent-close is when we're doing recovery and expect that a later call will reconstitute/reopen the link. So if anyone is in the middle of a retry loop the next call will open it back up.

Close permanently allows us (and other callers) to know that retries should end - we're in a locally non-retriable state. This will exit any retry loops.

Right now the way the code is written is that each caller is basically doing a liveness check when they get the client or use the client. This is just an extension of that logic.

Reply:
Makes sense given that we do lazy recreation.